File: middleware-node/src/authz.test.js
import request from 'supertest';
import express from 'express';
import { useAuthZ } from './authz.js';

const app = express();

describe('useAuthZ', () => {
  describe('Tests different user role values', () => {
    let user = {
      uid: 'UID123',
      email: 'x@gmail.com',
      role: '',
    };

    beforeEach(() => {
      app.use((req, res, next) => {
        res.locals = { user };
        next();
      });
    });

    const testCases = [
      {
        testName: 'Should succeed as user role is admin',
        role: 'admin',
        requiredRole: 'admin',
        expected: 200,
      },
      {
        testName: 'Should fail as user role is user but required role is admin',
        role: 'user',
        requiredRole: 'admin',
        expected: 403,
      },
      {
        testName: 'Should fail as user role is empty string',
        role: '',
        requiredRole: 'admin',
        expected: 401,
      },
      {
        testName: 'Should fail as user is not set',
        requiredRole : 'admin',
        expected: 401,
      },
    ];

    it.each(testCases)('%s', async (testCase) => {
      const { role, requiredRole, expected } = testCase;
      user.role = role;

      app.get('/', useAuthZ(requiredRole), (req, res) => {
        res.send('Success');
      });

      const response = await request(app).get('/');
      expect(response.statusCode).toBe(expected);
    });
  });

  it('Should fail as user is not present in context', async () => {
    app.get('/', useAuthZ('admin'), (req, res) => {
      res.send('Success');
    });

    const response = await request(app).get('/');
    expect(response.statusCode).toBe(401);
  });
});

File: middleware-go/src/middlewares/authz_test.go
package middlewares

import (
	"net/http/httptest"
	"testing"

	m "middleware-go/src/models"

	"github.com/gofiber/fiber/v2"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestAuthZMiddleware(t *testing.T) {
    tests := []struct {
        role string
        requiredRole string
        testName string
        expected int
    }{
        {
            role: "user",
            requiredRole: "admin",
            testName: "Should fail as user role is not admin",
            expected: 403,
        },
        {
            role: "",
            requiredRole: "admin",
            testName: "Should fail as user role is not set",
            expected: 401,
        },
        {
            role: "admin",
            requiredRole: "admin",
            testName: "Should pass as user role is admin",
            expected: 200,
        },
    }

    for _, test := range tests {
        t.Run(test.testName, func(t *testing.T) {
            app := fiber.New()
            // adding user to context as we test only the authorization part
            user := &m.User{
                UID: "UID123",
                Email: "x@gmail.com",
                Role: test.role,
            }
            app.Use(func(c *fiber.Ctx) error {
                c.Locals("user", user)
                return c.Next()
            })

            app.Get("/", UseAuthZ(test.requiredRole), func(c *fiber.Ctx) error {
                return c.SendString("Success")
            })
            req := httptest.NewRequest("GET", "/", nil)

            resp, err := app.Test(req)
            require.NoError(t, err)
            assert.Equal(t, test.expected, resp.StatusCode)
        })
    }

    t.Run("Should fail as user is not present in context", func(t *testing.T) {
        app := fiber.New()
        app.Get("/", UseAuthZ("admin"), func(c *fiber.Ctx) error {
            return c.SendString("Success")
        })
        req := httptest.NewRequest("GET", "/", nil)

        resp, err := app.Test(req)
        require.NoError(t, err)
        assert.Equal(t, 401, resp.StatusCode)
    })
}
